src/io.rs

```diff
@@ -97,4 +99,85 @@ impl Write for CallbackWriter {
     fn flush(&mut self) -> Result<()> {
         Ok(())
     }
+
+    fn write_vectored(&mut self, bufs: &[IoSlice<'_>]) -> Result<usize> {
+        let mut out_n: usize = 0;
+        for buf in bufs.iter().filter(|b| !b.is_empty()) {
+            out_n += match self.write(buf) {
+                Ok(n) => {
+                    if n != buf.len() {
+                        return Ok(out_n + n);
+                    }
+                    n
+                }
+                e => return e,
+            }
+        }
+        Ok(out_n)
+    }
+}
+
+/// A callback for rustls_connection_write_tls_vectored.
+/// An implementation of this callback should attempt to write the bytes in
+/// the given `count` rustls_slice_bytes to the network. If any bytes were written,
+/// the implementation should set out_n to the number of bytes written and return 0.
+/// If there was an error, the implementation should return a nonzero rustls_io_result,
+/// which will be passed through to the caller. On POSIX systems, returning `errno` is convenient.
+/// On other systems, any appropriate error code works.
+/// (including EAGAIN or EWOULDBLOCK), the implementation should return `errno`.
+/// It's best to make one write attempt to the network per call. Additional write will
+/// be triggered by subsequent calls to one of the `_write_tls` methods.
+/// `userdata` is set to the value provided to `rustls_*_session_set_userdata`. In most
+/// cases that should be a struct that contains, at a minimum, a file descriptor.
+/// The buf and out_n pointers are borrowed and should not be retained across calls.
+pub type rustls_write_vectored_callback = Option<
+    unsafe extern "C" fn(
+        userdata: *mut c_void,
+        slices: *const rustls_slice_bytes,
+        count: size_t,
+        out_n: *mut size_t,
+    ) -> rustls_io_result,
+>;
+
+pub(crate) type VectoredWriteCallback = unsafe extern "C" fn(
+    userdata: *mut c_void,
+    slices: *const rustls_slice_bytes,
+    count: size_t,
+    out_n: *mut size_t,
+) -> rustls_io_result;
+
+pub(crate) struct VectoredCallbackWriter {
+    pub callback: VectoredWriteCallback,
+    pub userdata: *mut c_void,
+}
+
+impl Write for VectoredCallbackWriter {
+    fn write(&mut self, buf: &[u8]) -> Result<usize> {
+        let mut out_n: usize = 0;
+        let cb = self.callback;
+        let slices: &[rustls_slice_bytes] = &[rustls_slice_bytes::from(buf)];
+        let result = unsafe { cb(self.userdata, slices.as_ptr(), slices.len(), &mut out_n) };
+        match result.0 {
+            0 => Ok(out_n),
+            e => Err(Error::from_raw_os_error(e)),
+        }
+    }
+
+    fn flush(&mut self) -> Result<()> {
+        Ok(())
+    }
+
+    fn write_vectored(&mut self, bufs: &[IoSlice<'_>]) -> Result<usize> {
+        let mut out_n: usize = 0;
+        let cb = self.callback;
+        let mut slices: Vec<rustls_slice_bytes> = Vec::new();
+        for buf in bufs.iter().filter(|b| !b.is_empty()) {
```

Comment:
Do we want to provide a guarantee, separate from rustls, that we never pass zero-length slices to the callback? If so, why. It seems like rustls probably doesn't create such slices under normal circumstances. If we need additional guarantees for C code we should make sure to document them.